### ISSUE TITLE
force uninstall

### ISSUE THREAD

**Issue by pohly** [operator-framework/operator-sdk#5451]
## Feature Request

Enable `operator-sdk olm uninstall` to remove partial OLM installations.

#### Describe the problem you need a feature to resolve.

operator-sdk: 1.6.1
OLM: 0.18.3

We install and uninstall OLM in the PMEM-CSI CI. Sometimes when OLM installation fails, the test cluster is left in a state where neither `operator-sdk olm install` nor `operator-sdk olm uninstall` work:

```
$ operator-sdk olm install --version=v0.18.3 --verbose --timeout=5m
    time="2021-12-11T22:36:51Z" level=debug msg="Debug logging is set"
    time="2021-12-11T22:36:51Z" level=info msg="Fetching CRDs for version \"v0.18.3\""
    time="2021-12-11T22:36:51Z" level=info msg="Fetching resources for resolved version \"v0.18.3\""
    time="2021-12-11T22:36:52Z" level=fatal msg="Failed to install OLM version \"v0.18.3\": detected existing OLM resources: OLM must be completely uninstalled before installation"

$ operator-sdk olm uninstall
    time="2021-12-11T22:36:53Z" level=fatal msg="Failed to uninstall OLM: error getting installed OLM version (set --version to override the default version): no existing installation found"
```

#### Describe the solution you'd like.

A "force install" for incomplete installations would help resolve this situation.

Alternatively, provide more information about which "existing OLM resources" were detected so that someone can manually remove them. I already tried to delete everything defined by https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml and that has helped in some cases, but not all. In this case nothing from that YAML file seems to be left but `operator-sdk olm install` still failed:

https://cloudnative-k8sci.southcentralus.cloudapp.azure.com/view/pmem-csi/job/pmem-csi-periodic/1406/testReport/junit/fedora-1_19.olm-lvm-production/raw-conversion/Testing___1_19___works/

```
time="2021-12-11T22:36:48Z" level=fatal msg="Failed to uninstall OLM: error getting installed OLM version (set --version to override the default version): no existing installation found"
    + /mnt/workspace/pmem-csi-periodic/_work/govm/ssh.0 kubectl delete -f https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": namespaces "olm" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": namespaces "operators" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": serviceaccounts "olm-operator-serviceaccount" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": clusterroles.rbac.authorization.k8s.io "system:controller:operator-lifecycle-manager" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": clusterrolebindings.rbac.authorization.k8s.io "olm-operator-binding-olm" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": deployments.apps "olm-operator" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": deployments.apps "catalog-operator" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": clusterroles.rbac.authorization.k8s.io "aggregate-olm-edit" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": clusterroles.rbac.authorization.k8s.io "aggregate-olm-view" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": operatorgroups.operators.coreos.com "global-operators" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": operatorgroups.operators.coreos.com "olm-operators" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": clusterserviceversions.operators.coreos.com "packageserver" not found
    Error from server (NotFound): error when deleting "https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/olm.yaml": catalogsources.operators.coreos.com "operatorhubio-catalog" not found
    + /mnt/workspace/pmem-csi-periodic/_work/govm/ssh.0 kubectl delete ns olm
    Error from server (NotFound): namespaces "olm" not found
    + true
    + /mnt/workspace/pmem-csi-periodic/_work/bin/operator-sdk olm install --version=v0.18.3 --verbose --timeout=5m
    time="2021-12-11T22:36:51Z" level=debug msg="Debug logging is set"
    time="2021-12-11T22:36:51Z" level=info msg="Fetching CRDs for version \"v0.18.3\""
    time="2021-12-11T22:36:51Z" level=info msg="Fetching resources for resolved version \"v0.18.3\""
    time="2021-12-11T22:36:52Z" level=fatal msg="Failed to install OLM version \"v0.18.3\": detected existing OLM resources: OLM must be completely uninstalled before installation"
```

**Comment by jberkhahn** [comment]
Hey there,

We tried and were unable to reproduce this. We noticed you're on a pretty old version of Operator SDK, and we were doing this with the newest (1.17). Could you try this with a newer version and see if it works now? 

If you're able to reproduce this bug, could you give us the exact sequence of steps you took to do so, including both versions of OLM you installed?

Thanks

**Comment by pohly** [comment]
I cannot reproduce this reliably either.

From experience I know that the cluster can get into this broken state when installing the OLM CRDs worked, but then creating CRs failed. Does "operator-sdk olm install" retry creating CRs? I think I have seen cases where a CRD was created, but an immediate attempt to use them fails. My theory is that the apiserver still needs some time to process the new CRDs.

Another possibility is that downloading "crds.yaml" worked, but then downloading "olm.yaml" didn't.

So one way to reproduce this would be to run "operator-sdk olm install" under a debugger and kill it after the "crds.yaml" was created. Or do this manually:
```
$ kubectl create -f https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.20.0/crds.yaml
customresourcedefinition.apiextensions.k8s.io/catalogsources.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/clusterserviceversions.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/installplans.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/olmconfigs.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/operatorconditions.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/operatorgroups.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/operators.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/subscriptions.operators.coreos.com created
$ _work/bin/operator-sdk olm install --version=v0.20.0
INFO[0000] Fetching CRDs for version "v0.20.0"          
INFO[0000] Fetching resources for resolved version "v0.20.0" 
FATA[0001] Failed to install OLM version "v0.20.0": detected existing OLM resources: OLM must be completely uninstalled before installation 
$ _work/bin/operator-sdk olm uninstall
FATA[0000] Failed to uninstall OLM: error getting installed OLM version (set --version to override the default version): no existing installation found 
```

**Comment by jberkhahn** [comment]
WIth no way to reliably reproduce this we can't really fix it. Closing this, but feel free to reopen if you can reliably reproduce the bug.
